File: packages/hush-ts/src/guards/forbidden-path.ts
import { Guard, GuardAction, GuardContext, GuardResult, Severity } from "./types";

const DEFAULT_FORBIDDEN_PATTERNS = [
  // SSH keys
  "**/.ssh/**",
  "**/id_rsa*",
  "**/id_ed25519*",
  "**/id_ecdsa*",
  // AWS credentials
  "**/.aws/**",
  // Environment files
  "**/.env",
  "**/.env.*",
  // Git credentials
  "**/.git-credentials",
  "**/.gitconfig",
  // GPG keys
  "**/.gnupg/**",
  // Kubernetes
  "**/.kube/**",
  // Docker
  "**/.docker/**",
  // NPM tokens
  "**/.npmrc",
  // Password stores
  "**/.password-store/**",
  "**/pass/**",
  // 1Password
  "**/.1password/**",
  // System paths
  "/etc/shadow",
  "/etc/passwd",
  "/etc/sudoers",
];

export interface ForbiddenPathConfig {
  patterns?: string[];
  exceptions?: string[];
}

/**
 * Guard that blocks access to sensitive paths.
 */
export class ForbiddenPathGuard implements Guard {
  readonly name = "forbidden_path";
  private patterns: string[];
  private exceptions: string[];

  constructor(config: ForbiddenPathConfig = {}) {
    this.patterns = config.patterns ?? DEFAULT_FORBIDDEN_PATTERNS;
    this.exceptions = config.exceptions ?? [];
  }

  handles(action: GuardAction): boolean {
    return ["file_access", "file_write", "patch"].includes(action.actionType);
  }

  check(action: GuardAction, _context: GuardContext): GuardResult {
    if (!this.handles(action)) {
      return GuardResult.allow(this.name);
    }

    const path = action.path;
    if (!path) {
      return GuardResult.allow(this.name);
    }

    if (this.isForbidden(path)) {
      return GuardResult.block(
        this.name,
        Severity.CRITICAL,
        `Access to forbidden path: ${path}`
      ).withDetails({
        path,
        reason: "matches_forbidden_pattern",
      });
    }

    return GuardResult.allow(this.name);
  }

  private isForbidden(path: string): boolean {
    // Normalize path (handle Windows paths)
    const normalized = path.replace(/\\/g, "/");

    // Check exceptions first
    for (const exception of this.exceptions) {
      if (matchGlob(normalized, exception)) {
        return false;
      }
    }

    // Check forbidden patterns
    for (const pattern of this.patterns) {
      if (matchGlob(normalized, pattern)) {
        return true;
      }
    }

    return false;
  }
}

/**
 * Simple glob matcher supporting:
 * - * matches any characters except /
 * - ** matches any characters including /
 * - ? matches any single character
 */
function matchGlob(path: string, pattern: string): boolean {
  // Convert glob to regex
  let regex = pattern
    .replace(/\*\*/g, "\u0000") // Placeholder for **
    .replace(/\*/g, "[^/]*")
    .replace(/\u0000/g, ".*")
    .replace(/\?/g, ".");

  // Escape dots that aren't part of patterns
  regex = regex.replace(/\.(?!\*)/g, "\\.");

  // Anchor the pattern
  if (!regex.startsWith(".*") && !regex.startsWith("/")) {
    regex = "(^|.*/)" + regex;
  }
  regex = "^" + regex + "$";

  return new RegExp(regex).test(path);
}

File: packages/sdk/hush-py/src/clawdstrike/guards/forbidden_path.py
"""Forbidden path guard - blocks access to sensitive paths."""

from __future__ import annotations

from dataclasses import dataclass, field

from wcmatch import glob as wcglob

from clawdstrike.guards.base import (
    Action,
    Guard,
    GuardContext,
    GuardResult,
    Severity,
)


def default_forbidden_patterns() -> list[str]:
    """Default patterns for forbidden paths."""
    return [
        # SSH keys
        "**/.ssh/**",
        "**/id_rsa*",
        "**/id_ed25519*",
        "**/id_ecdsa*",
        # AWS credentials
        "**/.aws/**",
        # Environment files
        "**/.env",
        "**/.env.*",
        # Git credentials
        "**/.git-credentials",
        "**/.gitconfig",
        # GPG keys
        "**/.gnupg/**",
        # Kubernetes
        "**/.kube/**",
        # Docker
        "**/.docker/**",
        # NPM tokens
        "**/.npmrc",
        # Password stores
        "**/.password-store/**",
        "**/pass/**",
        # 1Password
        "**/.1password/**",
        # System paths
        "/etc/shadow",
        "/etc/passwd",
        "/etc/sudoers",
    ]


@dataclass
class ForbiddenPathConfig:
    """Configuration for ForbiddenPathGuard."""

    patterns: list[str] = field(default_factory=default_forbidden_patterns)
    exceptions: list[str] = field(default_factory=list)


class ForbiddenPathGuard(Guard):
    """Guard that blocks access to sensitive paths."""

    def __init__(self, config: ForbiddenPathConfig | None = None) -> None:
        self._config = config or ForbiddenPathConfig()

    @property
    def name(self) -> str:
        return "forbidden_path"

    def handles(self, action: Action) -> bool:
        return action.action_type in ("file_access", "file_write", "patch")

    def is_forbidden(self, path: str) -> bool:
        """Check if a path is forbidden.

        Args:
            path: Path to check

        Returns:
            True if path is forbidden, False otherwise
        """
        # Normalize path (handle Windows paths)
        normalized = path.replace("\\", "/")

        # Check exceptions first
        for exception in self._config.exceptions:
            if wcglob.globmatch(normalized, exception, flags=wcglob.GLOBSTAR):
                return False

        # Check forbidden patterns
        for pattern in self._config.patterns:
            if wcglob.globmatch(normalized, pattern, flags=wcglob.GLOBSTAR):
                return True

        return False

    def check(self, action: Action, context: GuardContext) -> GuardResult:
        """Check if file access is allowed.

        Args:
            action: The action to check
            context: Execution context

        Returns:
            GuardResult
        """
        if not self.handles(action):
            return GuardResult.allow(self.name)

        path: str | None = getattr(action, "path", None)
        if path is None:
            return GuardResult.allow(self.name)

        if self.is_forbidden(path):
            return GuardResult.block(
                self.name,
                Severity.CRITICAL,
                f"Access to forbidden path: {path}",
            ).with_details({
                "path": path,
                "reason": "matches_forbidden_pattern",
            })

        return GuardResult.allow(self.name)


__all__ = ["ForbiddenPathGuard", "ForbiddenPathConfig"]
